pydantic/dataclasses.py

```diff
@@ -264,6 +265,35 @@ def create_dataclass(cls: type[Any]) -> type[PydanticDataclass]:
             **kwargs,
         )
 
+        if config_wrapper.validate_assignment:
+
+            @functools.wraps(cls.__setattr__)
+            def validated_setattr(instance: Any, field: str, value: str, /) -> None:
+                type(instance).__pydantic_validator__.validate_assignment(instance, field, value)
+
+            cls.__setattr__ = validated_setattr.__get__(None, cls)  # type: ignore
+
+            if slots and not hasattr(cls, '__setstate__'):
+                # If slots is set, `pickle` (relied on by `copy.copy()`) will use
+                # `__setattr__()` to reconstruct the dataclass. However, the custom
+                # `__setattr__()` set above relies on `validate_assignment()`, which
+                # in turn excepts all the field values to be already present on the
+                # instance, resulting in attribute errors.
+                # As such, we make use of `object.__setattr__()` instead.
+                # Note that we do so only if `__setstate__()` isn't already set (this is the
+                # case if on top of `slots`, `frozen` is used).
```

Comment:
Is it required to skip this when frozen is set on top of slots? https://github.com/python/cpython/pull/104041#issue-1690829773 suggests a bug was fixed hereabouts.

Reply:
When a frozen and slotted dataclass is created, the stdlib adds a `__set/getstate__()` implementation (and we actually reuse this implementation below). We could also unconditionally set our own `__set/getstate__()`, but this would also override any user-defined one (the CPython PR you linked is actually a fix for this on their end)?